addons/package_hierarchy/models/stock_quant_package.py

```diff
@@ -0,0 +1,222 @@
+        "children_ids.quant_ids.package_id",
+    )
+    def _compute_children_quant_ids(self):
+        for package in self:
+            if isinstance(package.id, models.NewId):
+                package.children_quant_ids = []
+            else:
+                package._check_not_multi_location()
+                package.children_quant_ids = self.env["stock.quant"].search(
+                    [
+                        ("package_id", "child_of", package.id),
+                        "|",
+                        ("quantity", "!=", 0),
+                        ("reserved_quantity", "!=", 0),
+                    ]
+                )
+
+    @api.depends(
+        "quant_ids.package_id",
+        "quant_ids.location_id",
+        "quant_ids.company_id",
+        "quant_ids.owner_id",
+        "quant_ids.quantity",
+        "quant_ids.reserved_quantity",
+    )
+    def _compute_package_info(self):
+        for package in self:
+            values = {
+                "location_id": False,
+                "owner_id": False,
+            }
+
+            comparison_recs = False
+
+            if package.quant_ids:
+                comparison_recs = package.quant_ids
+            elif package.children_ids:
+                comparison_recs = package.children_ids
+
+            if comparison_recs:
+                values["location_id"] = comparison_recs[0].location_id
+                if all(q.owner_id == comparison_recs[0].owner_id for q in comparison_recs):
+                    values["owner_id"] = comparison_recs[0].owner_id
+                if all(q.company_id == comparison_recs[0].company_id for q in comparison_recs):
+                    values["company_id"] = comparison_recs[0].company_id
+
+            package.location_id = values["location_id"]
+            package.company_id = values.get("company_id")
+            package.owner_id = values["owner_id"]
+
+    @api.depends("parent_id")
+    def _compute_display_name(self):
+        """Compute the display name for a package. Include name of immediate parent."""
+        for package in self:
+            if package.parent_id:
+                package.display_name = "%s/%s" % (package.parent_id.name, package.name)
+            else:
+                package.display_name = package.name
+
+    def _get_contained_quants(self):
+        """Overide to include picks quants of child packages"""
+        return self.env["stock.quant"].search([("package_id", "child_of", self.ids)])
+
+    def _get_move_lines_of_children_domain(self):
+        return [
+            "|",
+            ("result_package_id", "child_of", self.ids),
+            ("package_id", "child_of", self.ids),
+        ]
+
+    def get_move_lines_of_children(self, aux_domain=None, **kwargs):
+        MoveLines = self.env["stock.move.line"]
+        domain = self._get_move_lines_of_children_domain()
+        if aux_domain:
+            domain.extend(aux_domain)
+        return MoveLines.search(domain, **kwargs)
```

Comment:
Unless we are interested on keeping the default ordering, we could add order='id' to avoid a join with stock.quant.package table.